### PR TITLE
feat: tokens

### LINKED ISSUE #3692: Create long-lived tokens for CI/CD/automation (or tokens for a specific duration)
Add a way (likely via the CLI) to generate API tokens:

- push templates via CI
- use the REST API/CLI in automated pipelines

## Original description

I searched the issues with multiple keywords but I could not find any information on how to push template updates from Github Actions for example. Is that actually possible? So far I tried the username/password approach but that fails because it requires the user to enter the token from the browser.

### PR DESCRIPTION
**Pull request by f0ssel**
Closes https://github.com/coder/coder/issues/3692

### CODE REVIEW

File: cli/keys.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+func keys() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "keys",
+		Short:   "Manage machine keys",
+		Long:    "Machine keys are used to authenticate automated clients to Coder.",
+		Aliases: []string{"key"},
+		Example: formatExamples(
+			example{
+				Description: "Create a machine key for CI/CD scripts",
```
Comment (kylecarbs) [inline cli/keys.go:23]:
```suggestion
				Description: "Create a machine key for automation",
```

File: cli/keys.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+			}
+
+			cmd.Println(cliui.Styles.Wrap.Render(
+				"Here is your API key. 🪄",
+			))
+			cmd.Println()
+			cmd.Println(cliui.Styles.Code.Render(strings.TrimSpace(res.Key)))
+			cmd.Println()
+			cmd.Println(cliui.Styles.Wrap.Render(
+				fmt.Sprintf("You can use this API key by setting --%s CLI flag, the %s environment variable, or the %q HTTP header.", varToken, envSessionToken, codersdk.SessionTokenKey),
```
Comment (kylecarbs) [inline cli/keys.go:67]:
```suggestion
				fmt.Sprintf("You can use this API key by setting the --%s CLI flag, %s environment variable, or the %q HTTP header.", varToken, envSessionToken, codersdk.SessionTokenKey),
```

File: coderd/coderd.go [hunk 3]
```diff
@@ -399,8 +399,14 @@ func New(options *Options) *API {
 					r.Get("/roles", api.userRoles)
 
 					r.Route("/keys", func(r chi.Router) {
-						r.Post("/", api.postAPIKey)
```
Comment (kylecarbs) [inline coderd/coderd.go:402]: This route was used on the FE for CLI auth. We'll have to change it there too. Otherwise, this will break it.

**Review by bpmct — COMMENTED** [review]
Are "machine key" and "personal access token" interchangeable?

I prefer PAT / "token", but don't have strong feelings either way, especially if the behaviors are the same. The main rationale is I haven't seen "machine key" used commonly in software like this, but 🤷🏼

**Comment by coadler** [comment]
I find "machine key" a bit confusing as we have machine keys in Tailscale as well.

File: coderd/coderd.go (hunk 3 above)
Comment (f0ssel) [reply coderd/coderd.go:402]: I did not find anything calling this route from our api.ts or codersdk, do you know of somewhere else?

**Comment by f0ssel** [comment]
machine key is programmer codename 100% - it's because the data model is an api key with the login type "machine". I can change this to whatever makes the most sense for product. 

In v1 we used `coder tokens` and also call the api key flag in the CLI for v2 `--token`. Can we call it just `coder tokens` again? PAT doesn't really mean anything to me personally.

**Comment by bpmct** [comment]
Tokens makes sense to me!

File: coderd/apikey.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+func (api *API) postToken(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	user := httpmw.UserParam(r)
+
+	if !api.Authorize(r, rbac.ActionCreate, rbac.ResourceAPIKey.WithOwner(user.ID.String())) {
+		httpapi.ResourceNotFound(rw)
+		return
+	}
+
+	// machine keys last 100 years
```
Comment (deansheather) [inline coderd/apikey.go:37]:
```suggestion
	// tokens last 100 years
```

File: coderd/users_test.go [hunk 5]
```diff
@@ -285,15 +285,14 @@ func TestPostLogin(t *testing.T) {
 		require.NoError(t, err, "fetch login key")
 		require.Equal(t, int64(86400), key.LifetimeSeconds, "default should be 86400")
 
-		// Generated tokens have a longer life
-		token, err := client.CreateAPIKey(ctx, admin.UserID.String())
-		require.NoError(t, err, "make new api key")
+		// Machine tokens have a longer life
```
Comment (deansheather) [inline coderd/users_test.go:288]:
```suggestion
		// Generated tokens have a longer life
```

File: coderd/users_test.go [hunk 6]
```diff
@@ -1195,36 +1194,18 @@ func TestGetUsers(t *testing.T) {
 	})
 }
 
-func TestPostAPIKey(t *testing.T) {
+func TestPostMachineKey(t *testing.T) {
```
Comment (deansheather) [inline coderd/users_test.go]:
```suggestion
func TestPostTokens(t *testing.T) {
```

**Review by deansheather — APPROVED** [review]
Looks fine but you still say "machine token" in a lot of places. Also, using both "api key" and "token" and "key" and "session token" for the same thing is confusing and we should just stick to one where possible IMO

**Pull request merged**